File: android/app/src/main/java/com/stuartmcmullen/salonproject/StripeTapToPayPlugin.java
package com.stuartmcmullen.salonproject;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.stripe.stripeterminal.Terminal;
import com.stripe.stripeterminal.external.callable.Callback;
import com.stripe.stripeterminal.external.callable.ConnectionTokenCallback;
import com.stripe.stripeterminal.external.callable.ConnectionTokenProvider;
import com.stripe.stripeterminal.external.callable.DiscoveryListener;
import com.stripe.stripeterminal.external.callable.PaymentIntentCallback;
import com.stripe.stripeterminal.external.callable.ReaderCallback;
import com.stripe.stripeterminal.external.callable.TapToPayReaderListener;
import com.stripe.stripeterminal.external.callable.TerminalListener;
import com.stripe.stripeterminal.external.models.ConnectionConfiguration;
import com.stripe.stripeterminal.external.models.DisconnectReason;
import com.stripe.stripeterminal.external.models.DiscoveryConfiguration;
import com.stripe.stripeterminal.external.models.PaymentIntent;
import com.stripe.stripeterminal.external.models.Reader;
import com.stripe.stripeterminal.external.models.TerminalException;
import com.stripe.stripeterminal.log.LogLevel;

import org.json.JSONArray;

import java.util.ArrayList;
import java.util.List;

@CapacitorPlugin(name = "StripeTapToPay")
public class StripeTapToPayPlugin extends Plugin implements ConnectionTokenProvider, DiscoveryListener, TapToPayReaderListener {
    private static final String TAG = "StripeTapToPay";

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private ConnectionTokenCallback connectionTokenCallback;
    private final List<Reader> discoveredReaders = new ArrayList<>();
    private PluginCall pendingDiscoverCall;
    private PaymentIntent currentPaymentIntent;
    private String lastLocationId;

    @PluginMethod
    public void initialize(PluginCall call) {
        Log.d(TAG, "initialize() called");
        mainHandler.post(() -> {
            try {
                if (!Terminal.isInitialized()) {
                    Terminal.init(
                        getContext().getApplicationContext(),
                        LogLevel.VERBOSE,
                        this,
                        new TerminalListener() {},
                        null
                    );
                }
                call.resolve();
            } catch (Exception e) {
                call.reject(e.getMessage(), e);
            }
        });
    }

    @Override
    public void fetchConnectionToken(@NonNull ConnectionTokenCallback callback) {
        if (connectionTokenCallback != null) {
            Log.d(TAG, "Ignoring duplicate token request while one is already pending");
            return;
        }
        Log.d(TAG, "fetchConnectionToken requested by SDK");
        connectionTokenCallback = callback;
        notifyListeners("requestedConnectionToken", new JSObject());
    }

    @PluginMethod
    public void setConnectionToken(PluginCall call) {
        String token = call.getString("token");
        if (token == null || token.isEmpty()) {
            call.reject("Missing connection token");
            return;
        }

        if (connectionTokenCallback != null) {
            Log.d(TAG, "setConnectionToken received token, completing pending request");
            connectionTokenCallback.onSuccess(token);
            connectionTokenCallback = null;
        } else {
            Log.d(TAG, "setConnectionToken received token but no pending completion exists");
        }
        call.resolve();
    }

    @PluginMethod
    public void discoverReaders(PluginCall call) {
        mainHandler.post(() -> {
            try {
                if (!Terminal.isInitialized()) {
                    call.reject("Terminal not initialized");
                    return;
                }

                boolean isSimulated = Boolean.TRUE.equals(call.getBoolean("isSimulated", false));
                String locationId = call.getString("locationId");
                lastLocationId = locationId;
                pendingDiscoverCall = call;
                discoveredReaders.clear();

                DiscoveryConfiguration config = new DiscoveryConfiguration.TapToPayDiscoveryConfiguration(isSimulated);
                Terminal.getInstance().discoverReaders(config, this, new Callback() {
                    @Override
                    public void onSuccess() {
                        Log.d(TAG, "discoverReaders completed");
                        if (pendingDiscoverCall != null) {
                            JSObject result = new JSObject();
                            result.put("readers", serializeReaders(discoveredReaders));
                            pendingDiscoverCall.resolve(result);
                            pendingDiscoverCall = null;
                        }
                    }

                    @Override
                    public void onFailure(@NonNull TerminalException e) {
                        Log.e(TAG, "discoverReaders failed", e);
                        if (pendingDiscoverCall != null) {
                            pendingDiscoverCall.reject(e.getErrorMessage(), e);
                            pendingDiscoverCall = null;
                        }
                    }
                });
            } catch (Exception e) {
                call.reject(e.getMessage(), e);
            }
        });
    }

    @Override
    public void onUpdateDiscoveredReaders(@NonNull List<Reader> readers) {
        Log.d(TAG, "discoverReaders emitted " + readers.size() + " readers");
        discoveredReaders.clear();
        discoveredReaders.addAll(readers);
        if (pendingDiscoverCall != null) {
            JSObject result = new JSObject();
            result.put("readers", serializeReaders(readers));
            pendingDiscoverCall.resolve(result);
            pendingDiscoverCall = null;
        }
    }

    @PluginMethod
    public void connectReader(PluginCall call) {
        mainHandler.post(() -> {
            try {
                JSObject readerObj = call.getObject("reader");
                if (readerObj == null) {
                    call.reject("Missing reader");
                    return;
                }
                Reader reader = lookupReader(readerObj);
                if (reader == null) {
                    call.reject("Reader not found");
                    return;
                }
                String locationId = call.getString("locationId");
                if (locationId == null || locationId.isEmpty()) {
                    locationId = lastLocationId;
                }
                if (locationId == null || locationId.isEmpty()) {
                    call.reject("Missing locationId for Tap to Pay");
                    return;
                }

                ConnectionConfiguration.TapToPayConnectionConfiguration config = new ConnectionConfiguration.TapToPayConnectionConfiguration(locationId, this);
                Terminal.getInstance().connectReader(reader, config, new ReaderCallback() {
                    @Override
                    public void onSuccess(@NonNull Reader reader) {
                        JSObject result = new JSObject();
                        result.put("reader", serializeReader(reader));
                        call.resolve(result);
                    }

                    @Override
                    public void onFailure(@NonNull TerminalException e) {
                        call.reject(e.getErrorMessage(), e);
                    }
                });
            } catch (Exception e) {
                call.reject(e.getMessage(), e);
            }
        });
    }

    @PluginMethod
    public void disconnectReader(PluginCall call) {
        Terminal.getInstance().disconnectReader(new Callback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull TerminalException e) {
                call.reject(e.getErrorMessage(), e);
            }
        });
    }

    @PluginMethod
    public void collectPaymentMethod(PluginCall call) {
        String clientSecret = call.getString("paymentIntent");
        if (clientSecret == null || clientSecret.isEmpty()) {
            call.reject("Missing paymentIntent client secret");
            return;
        }

        Terminal.getInstance().retrievePaymentIntent(clientSecret, new PaymentIntentCallback() {
            @Override
            public void onSuccess(@NonNull PaymentIntent paymentIntent) {
                currentPaymentIntent = paymentIntent;
                Terminal.getInstance().collectPaymentMethod(paymentIntent, new PaymentIntentCallback() {
                    @Override
                    public void onSuccess(@NonNull PaymentIntent paymentIntent) {
                        currentPaymentIntent = paymentIntent;
                        JSObject result = new JSObject();
                        result.put("paymentIntent", serializePaymentIntent(paymentIntent));
                        call.resolve(result);
                    }

                    @Override
                    public void onFailure(@NonNull TerminalException e) {
                        call.reject(e.getErrorMessage(), e);
                    }
                });
            }

            @Override
            public void onFailure(@NonNull TerminalException e) {
                call.reject(e.getErrorMessage(), e);
            }
        });
    }

    @PluginMethod
    public void confirmPaymentIntent(PluginCall call) {
        if (currentPaymentIntent == null) {
            call.reject("No PaymentIntent available to confirm");
            return;
        }
        Terminal.getInstance().confirmPaymentIntent(currentPaymentIntent, new PaymentIntentCallback() {
            @Override
            public void onSuccess(@NonNull PaymentIntent paymentIntent) {
                currentPaymentIntent = paymentIntent;
                JSObject result = new JSObject();
                result.put("paymentIntent", serializePaymentIntent(paymentIntent));
                call.resolve(result);
            }

            @Override
            public void onFailure(@NonNull TerminalException e) {
                call.reject(e.getErrorMessage(), e);
            }
        });
    }

    @PluginMethod
    public void cancelPaymentIntent(PluginCall call) {
        if (currentPaymentIntent == null) {
            call.reject("No PaymentIntent available to cancel");
            return;
        }
        Terminal.getInstance().cancelPaymentIntent(currentPaymentIntent, new PaymentIntentCallback() {
            @Override
            public void onSuccess(@NonNull PaymentIntent paymentIntent) {
                currentPaymentIntent = paymentIntent;
                JSObject result = new JSObject();
                result.put("paymentIntent", serializePaymentIntent(paymentIntent));
                call.resolve(result);
            }

            @Override
            public void onFailure(@NonNull TerminalException e) {
                call.reject(e.getErrorMessage(), e);
            }
        });
    }

    @PluginMethod
    public void isDeviceCapable(PluginCall call) {
        JSObject result = new JSObject();
        boolean hasNfc = getContext().getPackageManager().hasSystemFeature(android.content.pm.PackageManager.FEATURE_NFC);
        result.put("value", hasNfc);
        if (!hasNfc) {
            result.put("error", "NFC not supported on this device");
        }
        call.resolve(result);
    }

    private Reader lookupReader(JSObject readerObj) {
        String serial = readerObj.getString("serialNumber");
        if (serial != null) {
            for (Reader reader : discoveredReaders) {
                if (serial.equals(reader.getSerialNumber())) {
                    return reader;
                }
            }
        }
        return discoveredReaders.isEmpty() ? null : discoveredReaders.get(0);
    }

    private JSONArray serializeReaders(List<Reader> readers) {
        JSONArray arr = new JSONArray();
        for (Reader reader : readers) {
            arr.put(serializeReader(reader));
        }
        return arr;
    }

    private JSObject serializeReader(Reader reader) {
        JSObject obj = new JSObject();
        obj.put("serialNumber", reader.getSerialNumber());
        obj.put("stripeId", reader.getId());
        obj.put("deviceType", "tap_to_pay");
        obj.put("locationId", reader.getLocation() != null ? reader.getLocation().getId() : lastLocationId);
        obj.put("label", reader.getLabel());
        obj.put("simulated", reader.isSimulated());
        return obj;
    }

    private JSObject serializePaymentIntent(PaymentIntent paymentIntent) {
        JSObject obj = new JSObject();
        obj.put("id", paymentIntent.getId());
        obj.put("clientSecret", paymentIntent.getClientSecret());
        obj.put("amount", paymentIntent.getAmount());
        obj.put("currency", paymentIntent.getCurrency());
        obj.put("livemode", paymentIntent.getLivemode());
        return obj;
    }

    @Override
    public void onDisconnect(@NonNull DisconnectReason reason) { }
}

File: ios/App/App/StripeTapToPayPlugin.swift
import Foundation
import Capacitor
import StripeTerminal

@available(iOS 15.0, *)
@objc public class StripeTapToPayPlugin: CAPPlugin, CAPBridgedPlugin, ConnectionTokenProvider, DiscoveryDelegate, TapToPayReaderDelegate {
    public let identifier = "StripeTapToPayPlugin"
    public let jsName = "StripeTapToPay"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setConnectionToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "discoverReaders", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connectReader", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnectReader", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "collectPaymentMethod", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "confirmPaymentIntent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelPaymentIntent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isDeviceCapable", returnType: CAPPluginReturnPromise)
    ]
    private var connectionTokenCompletion: ConnectionTokenCompletionBlock?
    private var discoverCancelable: Cancelable?
    private var pendingDiscoverCall: CAPPluginCall?
    private var discoveredReaders: [Reader] = []
    private var currentPaymentIntent: PaymentIntent?
    private var lastLocationId: String?

    @objc func initialize(_ call: CAPPluginCall) {
        CAPLog.print("⚡️ [StripeTapToPay] initialize() called")
        DispatchQueue.main.async {
            CAPLog.print("⚡️ [StripeTapToPay] initialize() running on main thread: \(Thread.isMainThread)")
            if !Terminal.isInitialized() {
                Terminal.setLogListener { logLine in
                    CAPLog.print("⚡️ [StripeTapToPay] \(logLine)")
                }
                // Stripe Terminal iOS calls are expected on the main thread.
                Terminal.initWithTokenProvider(self, delegate: nil, offlineDelegate: nil, logLevel: .verbose)
            }
            call.resolve()
        }
    }

    // MARK: - SCPConnectionTokenProvider
    public func fetchConnectionToken(_ completion: @escaping ConnectionTokenCompletionBlock) {
        DispatchQueue.main.async {
            if self.connectionTokenCompletion != nil {
                CAPLog.print("⚡️ [StripeTapToPay] Ignoring duplicate token request while one is already pending")
                return
            }
            CAPLog.print("⚡️ [StripeTapToPay] fetchConnectionToken requested by SDK")
            self.connectionTokenCompletion = completion
            self.notifyListeners("requestedConnectionToken", data: [:])
        }
    }

    @objc func setConnectionToken(_ call: CAPPluginCall) {
        guard let token = call.getString("token"), !token.isEmpty else {
            call.reject("Missing connection token")
            return
        }

        if let completion = connectionTokenCompletion {
            CAPLog.print("⚡️ [StripeTapToPay] setConnectionToken received token, completing pending request")
            completion(token, nil)
            connectionTokenCompletion = nil
        } else {
            CAPLog.print("⚡️ [StripeTapToPay] setConnectionToken received token but no pending completion exists")
        }
        call.resolve()
    }

    // MARK: - Discovery
    @objc func discoverReaders(_ call: CAPPluginCall) {
        CAPLog.print("⚡️ [StripeTapToPay] discoverReaders() called")
        DispatchQueue.main.async {
            CAPLog.print("⚡️ [StripeTapToPay] discoverReaders() running on main thread: \(Thread.isMainThread)")
            guard Terminal.isInitialized() else {
                call.reject("Terminal not initialized")
                return
            }

            if Terminal.shared.connectionStatus == .connected, let existingReader = Terminal.shared.connectedReader {
                CAPLog.print("⚡️ [StripeTapToPay] discoverReaders() reusing existing connected reader")
                call.resolve(["readers": [self.serializeReader(existingReader)], "alreadyConnected": true])
                return
            }

            let type = call.getString("type") ?? "tap-to-pay"
            let isSimulated = call.getBool("isSimulated") ?? false
            let locationId = call.getString("locationId")
            self.lastLocationId = locationId

            let config: DiscoveryConfiguration
            do {
                switch type {
                case "tap_to_pay", "tap-to-pay":
                    CAPLog.print("⚡️ [StripeTapToPay] Using TapToPay discovery (simulated=\(isSimulated))")
                    config = try TapToPayDiscoveryConfigurationBuilder()
                        .setSimulated(isSimulated)
                        .build()
                default:
                    CAPLog.print("⚡️ [StripeTapToPay] Using Internet discovery (simulated=\(isSimulated))")
                    let builder = InternetDiscoveryConfigurationBuilder()
                        .setSimulated(isSimulated)
                    if let loc = locationId {
                        _ = builder.setLocationId(loc)
                    }
                    config = try builder.build()
                }
            } catch {
                call.reject(error.localizedDescription)
                return
            }

            self.discoveredReaders = []
            self.pendingDiscoverCall = call

            if let existingCancelable = self.discoverCancelable {
                Task {
                    try? await existingCancelable.cancel()
                }
            }

            self.discoverCancelable = Terminal.shared.discoverReaders(config, delegate: self) { [weak self] error in
                guard let self = self else { return }
                DispatchQueue.main.async {
                    defer { self.discoverCancelable = nil }

                    if let error = error {
                        CAPLog.print("⚡️ [StripeTapToPay] discoverReaders failed: \(error.localizedDescription)")
                        self.pendingDiscoverCall?.reject(error.localizedDescription)
                        self.pendingDiscoverCall = nil
                        return
                    }

                    CAPLog.print("⚡️ [StripeTapToPay] discoverReaders completed")
                    if let pendingCall = self.pendingDiscoverCall {
                        let payload = self.discoveredReaders.map { self.serializeReader($0) }
                        pendingCall.resolve(["readers": payload])
                        self.pendingDiscoverCall = nil
                    }
                }
            }
        }
    }

    public func terminal(_ terminal: Terminal, didUpdateDiscoveredReaders readers: [Reader]) {
        DispatchQueue.main.async {
            CAPLog.print("⚡️ [StripeTapToPay] discoverReaders emitted \(readers.count) readers")
            self.discoveredReaders = readers

            if let pendingCall = self.pendingDiscoverCall {
                let payload = readers.map { self.serializeReader($0) }
                pendingCall.resolve(["readers": payload])
                self.pendingDiscoverCall = nil

                if let cancelable = self.discoverCancelable {
                    Task {
                        try? await cancelable.cancel()
                    }
                }
            }
        }
    }

    // MARK: - Connect / Disconnect
    @objc func connectReader(_ call: CAPPluginCall) {
        CAPLog.print("⚡️ [StripeTapToPay] connectReader() called")
        DispatchQueue.main.async {
            if Terminal.shared.connectionStatus == .connected, let existingReader = Terminal.shared.connectedReader {
                CAPLog.print("⚡️ [StripeTapToPay] connectReader() reusing existing connected reader")
                let payload = self.serializeReader(existingReader)
                call.resolve(["reader": payload, "alreadyConnected": true])
                return
            }

            guard let readerDict = call.getObject("reader") else {
                call.reject("Missing reader")
                return
            }

            guard let reader = self.lookupReader(from: readerDict) else {
                call.reject("Reader not found")
                return
            }

            let locationId = call.getString("locationId") ?? reader.locationId ?? self.lastLocationId ?? ""

            guard reader.deviceType == .tapToPay else {
                call.reject("Only Tap to Pay is supported on iOS")
                return
            }
            if locationId.isEmpty {
                call.reject("Missing locationId for Tap to Pay")
                return
            }

            let builder = TapToPayConnectionConfigurationBuilder(delegate: self, locationId: locationId)
            let connectionConfig: TapToPayConnectionConfiguration
            do {
                connectionConfig = try builder.build()
            } catch {
                call.reject(error.localizedDescription)
                return
            }

            Terminal.shared.connectReader(reader, connectionConfig: connectionConfig) { connectedReader, error in
                if let error = error {
                    call.reject(error.localizedDescription)
                    return
                }
                CAPLog.print("⚡️ [StripeTapToPay] connectReader() success")
                let payload = self.serializeReader(connectedReader ?? reader)
                call.resolve(["reader": payload])
            }
        }
    }

    @objc func disconnectReader(_ call: CAPPluginCall) {
        Terminal.shared.disconnectReader { error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            call.resolve()
        }
    }

    // MARK: - Payments
    @objc func collectPaymentMethod(_ call: CAPPluginCall) {
        CAPLog.print("⚡️ [StripeTapToPay] collectPaymentMethod() called")
        guard let clientSecret = call.getString("paymentIntent"), !clientSecret.isEmpty else {
            call.reject("Missing paymentIntent client secret")
            return
        }

        Terminal.shared.retrievePaymentIntent(clientSecret: clientSecret) { intent, error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            guard let intent = intent else {
                call.reject("PaymentIntent not found")
                return
            }
            self.currentPaymentIntent = intent

            _ = Terminal.shared.collectPaymentMethod(intent) { collectedIntent, collectError in
                if let collectError = collectError {
                    call.reject(collectError.localizedDescription)
                    return
                }
                if let collectedIntent = collectedIntent {
                    self.currentPaymentIntent = collectedIntent
                    call.resolve(["paymentIntent": self.serializePaymentIntent(collectedIntent)])
                } else {
                    call.reject("collectPaymentMethod failed")
                }
            }
        }
    }

    @objc func confirmPaymentIntent(_ call: CAPPluginCall) {
        CAPLog.print("⚡️ [StripeTapToPay] confirmPaymentIntent() called")
        guard let intent = currentPaymentIntent else {
            call.reject("No PaymentIntent available to confirm")
            return
        }

        _ = Terminal.shared.confirmPaymentIntent(intent) { confirmedIntent, error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            if let confirmedIntent = confirmedIntent {
                self.currentPaymentIntent = confirmedIntent
                call.resolve(["paymentIntent": self.serializePaymentIntent(confirmedIntent)])
            } else {
                call.reject("confirmPaymentIntent failed")
            }
        }
    }

    @objc func cancelPaymentIntent(_ call: CAPPluginCall) {
        guard let intent = currentPaymentIntent else {
            call.reject("No PaymentIntent available to cancel")
            return
        }

        Terminal.shared.cancelPaymentIntent(intent) { canceledIntent, error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            if let canceledIntent = canceledIntent {
                self.currentPaymentIntent = canceledIntent
                call.resolve(["paymentIntent": self.serializePaymentIntent(canceledIntent)])
            } else {
                call.reject("cancelPaymentIntent failed")
            }
        }
    }

    @objc func isDeviceCapable(_ call: CAPPluginCall) {
        CAPLog.print("⚡️ [StripeTapToPay] isDeviceCapable() called")
        let result = Terminal.shared.supportsReaders(of: .tapToPay, discoveryMethod: .tapToPay, simulated: false)
        switch result {
        case .success:
            call.resolve(["value": true])
        case .failure(let error):
            call.resolve(["value": false, "error": error.localizedDescription])
        }
    }

    // MARK: - Helpers
    private func lookupReader(from dict: [String: Any]) -> Reader? {
        if let serial = dict["serialNumber"] as? String {
            return discoveredReaders.first { $0.serialNumber == serial }
        }
        return discoveredReaders.first
    }

    private func serializeReader(_ reader: Reader) -> [String: Any] {
        return [
            "serialNumber": reader.serialNumber,
            "stripeId": reader.stripeId ?? "",
            "deviceType": deviceTypeString(reader.deviceType),
            "locationId": reader.locationId ?? "",
            "label": reader.label ?? "",
            "simulated": reader.simulated,
        ]
    }

    private func serializePaymentIntent(_ intent: PaymentIntent) -> [String: Any] {
        return [
            "id": intent.stripeId,
            "clientSecret": intent.clientSecret ?? "",
            "amount": intent.amount,
            "currency": intent.currency,
            "livemode": intent.livemode,
        ]
    }

    private func deviceTypeString(_ type: DeviceType) -> String {
        switch type {
        case .tapToPay: return "tap_to_pay"
        case .stripeS700, .stripeS700DevKit, .stripeS710, .stripeS710DevKit: return "business_reader"
        case .stripeM2, .wisePad3, .chipper2X: return "bluetooth"
        default: return "unknown"
        }
    }

    // MARK: - Required TapToPay reader delegate methods (no-op)
    public func tapToPayReader(_ reader: Reader, didStartInstallingUpdate update: ReaderSoftwareUpdate, cancelable: Cancelable?) {}
    public func tapToPayReader(_ reader: Reader, didReportReaderSoftwareUpdateProgress progress: Float) {}
    public func tapToPayReader(_ reader: Reader, didFinishInstallingUpdate update: ReaderSoftwareUpdate?, error: Error?) {}
    public func tapToPayReader(_ reader: Reader, didRequestReaderInput inputOptions: ReaderInputOptions) {}
    public func tapToPayReader(_ reader: Reader, didRequestReaderDisplayMessage displayMessage: ReaderDisplayMessage) {}
}
